### PR TITLE
Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}

### LINKED ISSUE #50584: Add a `vectorize` flag to torch.autograd.functional.{jacobian, hessian}
## 🚀 Feature

Add a `vectorize` flag to torch.autograd.functional.jacobian and torch.autograd.functional.hessian (default: False). Under the hood, the `vectorize` flag uses `vmap` as the backend to compute the jacobian and hessian, respectively, providing speedups to users.

e.g.

```py
import torch
x = torch.randn(5, requires_grad=True)
f = lambda x: x ** 2
expected = torch.autograd.functional.jacobian(f, x)
jac = torch.autograd.functional.jacobian(f, x, vectorize=True)
assert torch.allclose(jac, expected)
```

## Motivation

Jacobian computation (and by extension, hessian computation) in PyTorch today involves invoking `torch.autograd.grad` once per row of the jacobian. The following explains the procedure used by [torch.autograd.functional.jacobian](https://github.com/pytorch/pytorch/blob/00d432a1ed179eff52a9d86a0630f623bf20a37a/torch/autograd/functional.py#L432-L457) at a high-level:
- Our autograd engine computes vector-jacobian products without fully materializing a jacobian.
- To compute the first row of the jacobian, we use a vector-jacobian-product with (1, 0, 0, ...) as the vector.
- To compute the ith row of the jacobian, we use the ith unit vector.
- Finally, when we have all of the rows of the jacobian, we stack all of them together

Assuming a N by N jacobian, we need to invoke the autograd engine N times. The amount of overhead here (due to tensor creation and operator overhead) can be and is significant in a number of use cases like bayesian logistic regression.

## Alternatives

Instead of updating `jacobian` and `hessian`, we can expose `vmap` directly and tell users to use (pseudocode) `vmap(vjp)` to compute efficient jacobians. However, this would create a "trap" in our API where a user using `autograd.functional.jacobian` cannot benefit from these performance improvements.

## Additional context

Not all batching rules that are needed for jacobian and hessian computation are implemented: https://github.com/pytorch/pytorch/issues/49562. A good number of these may require writing new CUDA kernels from scratch. We'd like to offer the `vectorize=False` API so that we can begin speeding up user code without having them wait until we are finished with writing a substantial number of batching rules.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50915 Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}**
* #50744 Turn on batched grad testing for CriterionTest

Fixes #50584
Add a vectorize flag to torch.autograd.functional.jacobian and
torch.autograd.functional.hessian (default: False). Under the hood, the
vectorize flag uses vmap as the backend to compute the jacobian and
hessian, respectively, providing speedups to users.

Test Plan:
- I updated all of the jacobian and hessian tests to also use
vectorized=True
- I added some simple sanity check tests that check e.g. jacobian with
vectorized=False vs
jacobian with vectorized=True.
- The mechanism for vectorized=True goes through batched gradient
computation. We have separate tests for those (see other PRs in this
stack).

Differential Revision: [D26057674](https://our.internmc.facebook.com/intern/diff/D26057674)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -5812,7 +5836,62 @@ def foo(x, y):
...
+    def test_jacobian_vectorize_correctness_multi_input_multi_output(self):
+        def f(x, y):
+            return (x * x) @ y, x @ (x.sum(1) * y)
+
+        x = torch.randn(5, 3)
+        y = torch.randn(3, 5)
+        self._check_jacobian_vectorize_correctness(f, (x, y))
+
+    def _check_hessian_vectorize_correctness(self, f, inputs):
+        expected = autogradF.jacobian(f, inputs)
```
Comment (albanD) [inline test/test_autograd.py:5874]: These should be `hessian`?

File: test/test_autograd.py [hunk 2]
```diff
@@ -5827,30 +5906,36 @@ def bar3(a):
 
         inp = torch.rand(4)
         with self.assertRaisesRegex(TypeError, "The inputs given to hessian must be either a Tensor"):
-            res = autogradF.hessian(foo, (inp, 2))
+            res = autogradF.hessian(foo, (inp, 2), vectorize=True)
```
Comment (albanD) [inline test/test_autograd.py:5909]: Should these be `vectorize=vectorize` ?

File: test/test_autograd.py [hunk 3]
```diff
@@ -5864,43 +5949,55 @@ def bar2(a):
 
         inp = torch.rand(4)
         with self.assertRaisesRegex(RuntimeError, "Output 0 of the user-provided function does not require gradients."):
-            res = autogradF.hessian(foo, inp, strict=True)
-        res = autogradF.hessian(foo, inp, strict=False)
+            res = autogradF.hessian(foo, inp, strict=True, vectorize=True)
```
Comment (albanD) [inline test/test_autograd.py:5952]: same `vectorize=vectorize`

File: test/test_autograd.py [hunk 4]
```diff
@@ -5918,23 +6015,23 @@ def bad_reducer(x):
...
         self._assert_interleaved_struct(res, inputs, inputs)
 
-    def test_hessian_create_graph(self):
+    def _test_hessian_create_graph(self, vectorize):
         def pow_reducer(x):
             return x.pow(3).sum()
 
         inputs = torch.rand(2, 2, requires_grad=True)
-        res = autogradF.hessian(pow_reducer, inputs, create_graph=True)
+        res = autogradF.hessian(pow_reducer, inputs, create_graph=True, vectorize=True)
```
Comment (albanD) [inline test/test_autograd.py:6023]: same `vectorize=vectorize`?

File: torch/autograd/functional.py [hunk 5]
```diff
@@ -429,6 +436,42 @@ def jacobian(func, inputs, create_graph=False, strict=False):
     _check_requires_grad(outputs, "outputs", strict=strict)
 
     jacobian: Tuple[torch.Tensor, ...] = tuple()
+
+    if vectorize:
+        for outidx, out in enumerate(outputs):
```
Comment (albanD) [inline torch/autograd/functional.py:441]: contrary to what is said in the docstring, you actually call autograd.grad once per output here.
Is that expected?

File: torch/autograd/functional.py [hunk 6]
```diff
@@ -429,6 +436,42 @@ def jacobian(func, inputs, create_graph=False, strict=False):
...
+                    else:
+                        if strict:
+                            msg = (f"Output {outidx} of the user-provided function is "
+                                   f"independent of input {el_idx}. This is not allowed in "
+                                   "strict mode.")
+                            raise RuntimeError(msg)
+                        vj[el_idx] = torch.zeros_like(inputs[el_idx])
+                return tuple(vj)
+
+            basis_vectors = torch.eye(flat_out.numel(), dtype=flat_out.dtype, device=flat_out.device)
```
Comment (albanD) [inline torch/autograd/functional.py:464]: nit: we don't have a `eye_like` function?

File: torch/autograd/functional.py (hunk 5 above)
Comment (albanD) [reply torch/autograd/functional.py:441]: Can we actually do that in a single call by giving a tuple of output/grad_output?

File: torch/autograd/functional.py [hunk 7]
```diff
@@ -545,7 +595,7 @@ def jac_func(*inp):
         _check_requires_grad(jac, "jacobian", strict=strict)
         return jac
 
-    res = jacobian(jac_func, inputs, create_graph=create_graph, strict=strict)
+    res = jacobian(jac_func, inputs, create_graph=create_graph, strict=strict, vectorize=vectorize)
```
Comment (albanD) [inline torch/autograd/functional.py:598]: Should we also `vectorize` the call to jacobian in `jac_func` ?
Or that is not supported by vmap?

File: test/test_autograd.py (hunk 1 above)
Comment (zou3519) [reply test/test_autograd.py:5874]: Oops copy-paste error. Thanks for the catch!

File: torch/autograd/functional.py (hunk 6 above)
Comment (zou3519) [reply torch/autograd/functional.py:464]: Unfortunately we don't

File: torch/autograd/functional.py (hunk 7 above)
Comment (zou3519) [reply torch/autograd/functional.py:598]: The function that we are computing a hessian over only has 1 output with 1 element. Therefore the `jacobian` call in `jac_func` is equivalent to a `vjp` call and doesn't benefit from `vectorize=True`.

File: test/test_autograd.py (hunk 4 above)
Comment (albanD) [reply test/test_autograd.py:6023]: no worries!

File: torch/autograd/functional.py (hunk 7 above)
Comment (albanD) [reply torch/autograd/functional.py:598]: Ho right, I forgot that we have this constraint! Sounds good!

File: torch/autograd/functional.py (hunk 5 above)
Comment (zou3519) [reply torch/autograd/functional.py:441]: > contrary to what is said in the docstring, you actually call autograd.grad once per output here.
Is that expected?

For the docstring, I meant to say "... calls autograd.grad once per output".

> Can we actually do that in a single call by giving a tuple of output/grad_output?

I never thought about this before. It turns out that we can! (and probably should). Let me think about it for a bit and then update the PR

Comment (zou3519) [reply torch/autograd/functional.py:441]: So the answer is yes, we can and should just do one call. However, this comes with the following caveat:
- `vectorize=True` and `strict=True` doesn't quite work. This is because instead of returning `None` for autograd.grad, `vectorize=True` might cause it to return a tensor full of zeros.
- I am thinking of just raising an error message if the above two are passed in together.

Comment (albanD) [reply torch/autograd/functional.py:441]: I guess an error is ok for now.
But in the future, I think it would be ok to get just 0s for these and not raise the errror?

Comment (zou3519) [reply torch/autograd/functional.py:441]: > But in the future, I think it would be ok to get just 0s for these and not raise the errror?

Is it OK for `strict=False` to cause different behavior when `vectorize` is True and when it is False? I thought that would have been strange

Comment (albanD) [reply torch/autograd/functional.py:441]: You mean `strict=True` right?
Yes I agree with you that it is surprising and forbidding it is definitely the best move for now.

Comment (zou3519) [reply torch/autograd/functional.py]: Yes, I meant `strict=True`, my bad. That sounds good (y)

File: torch/autograd/functional.py [hunk 8]
```diff
@@ -364,7 +365,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
...
+    # in torch.autograd.functional.jacobian.
+    assert len(tensors) == len(tensor_numels)
+    assert len(tensors) > 0
+    total_numel = sum(tensor_numels)
+    diag_start_indices = (0, *torch.tensor(tensor_numels).cumsum(dim=0)[:-1].neg().unbind())
+    chunks = tuple(tensor.new_zeros(total_numel, tensor_numel)
+                   for tensor, tensor_numel in zip(tensors, tensor_numels))
+    for chunk, diag_start_idx in zip(chunks, diag_start_indices):
+        chunk.diagonal(diag_start_idx).fill_(1)
+    return chunks
```
Comment (zou3519) [inline torch/autograd/functional.py:395]: I'm not sure if there's a more elegant way of doing this. We could construct an NxN identity matrix and then split it into chunks and finally coerce them to the correct device and dtype, but that won't perform well if `tensors` are on different devices or have different dtypes because it would perform copies

File: torch/autograd/functional.py [hunk 9]
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
...
+        grad_outputs = _construct_standard_basis_for(outputs, output_numels)
+        flat_outputs = tuple(output.reshape(-1) for output in outputs)
+
+        # Step 2: Call vmap + autograd.grad
+        def vjp(grad_output):
+            vj = list(_autograd_grad(flat_outputs, inputs, grad_output, create_graph=create_graph))
+            for el_idx, vj_el in enumerate(vj):
+                if vj_el is not None:
+                    continue
+                vj[el_idx] = torch.zeros_like(inputs[el_idx])
```
Comment (albanD) [inline torch/autograd/functional.py:528]: nit: Shall we use `new_zeros` here as well?

File: torch/autograd/functional.py [hunk 10]
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
...
+            return tuple(vj)
+
+        jacobians_of_flat_output = vmap(vjp)(grad_outputs)
+
+        # Step 3: The returned jacobian is one big tensor per input. In this step,
+        # we split each Tensor by output.
+        jacobian = tuple(
+            tuple(jac_out.view(output.shape + inp.shape)
+                  for jac_out, output in zip(jac.split(output_numels, dim=0), outputs))
```
Comment (albanD) [inline torch/autograd/functional.py:537]: nit: I would find this more readable if it was not a one liner to be honest :D

File: torch/autograd/functional.py [hunk 11]
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
...
+                    continue
+                vj[el_idx] = torch.zeros_like(inputs[el_idx])
+            return tuple(vj)
+
+        jacobians_of_flat_output = vmap(vjp)(grad_outputs)
+
+        # Step 3: The returned jacobian is one big tensor per input. In this step,
+        # we split each Tensor by output.
+        jacobian = tuple(
+            tuple(jac_out.view(output.shape + inp.shape)
```
Comment (albanD) [inline torch/autograd/functional.py:536]: shouldn't the view be `view(output.shape, inp.shape)`?

File: torch/autograd/functional.py [hunk 12]
```diff
@@ -364,7 +365,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
...
+    # Precondition: tensors always has at least one element.
+    #
+    # See NOTE: [Computing jacobian with vmap and grad for multiple tensors]
+    # for context behind this function. All the pre-conditions are guarded for
+    # in torch.autograd.functional.jacobian.
+    assert len(tensors) == len(tensor_numels)
+    assert len(tensors) > 0
+    total_numel = sum(tensor_numels)
+    diag_start_indices = (0, *torch.tensor(tensor_numels).cumsum(dim=0)[:-1].neg().unbind())
+    chunks = tuple(tensor.new_zeros(total_numel, tensor_numel)
```
Comment (albanD) [inline torch/autograd/functional.py:391]: This made me realize that I think we're missing a small comment in the return type docstring: the Jacobian matrix will have the same dtype and device as its corresponding input.

Comment (zou3519) [reply torch/autograd/functional.py:391]: I can add one

File: torch/autograd/functional.py (hunk 11 above)
Comment (zou3519) [reply torch/autograd/functional.py:536]: There are two ways to use view:
- pass var-args ints: tensor.view(1, 2, 3, 4, 5)
- pass a list/tuple of ints: tensor.view([1, 2, 3, 4, 5])

`jac_out.view(output.shape + input.shape)` is the latter, but if we wanted to do the former we could do `view(*output.shape, *inp.shape)`

File: torch/autograd/functional.py (hunk 9 above)
Comment (zou3519) [reply torch/autograd/functional.py:528]: Is the suggestion `vj[el_idx] = vj[el_idx].new_zeros(inputs[el_idx].shape)`?

Comment (albanD) [reply torch/autograd/functional.py:528]: `vj[el_idx] = inputs[el_idx].new_zeros(inputs[el_idx].shape)` actually as `vj[el_idx]` is None here. But yes.
I started trying to use the `new_*` variants as much as possible to avoid future errors with vmap. But maybe that's not really necessary?

File: torch/autograd/functional.py (hunk 11 above)
Comment (albanD) [reply torch/autograd/functional.py:536]: Ho these are tuples, not numbers. My bad!

File: torch/autograd/functional.py (hunk 8 above)
Comment (albanD) [reply torch/autograd/functional.py:395]: I think this version is good.
And the comment to explain it is quite nice!
A single Tensor would increase the memory overhead quite a bit, especially when multiple devices/dtypes are involved.

File: torch/autograd/functional.py (hunk 10 above)
Comment (zou3519) [reply torch/autograd/functional.py:537]: It is more verbose if this were to be for loops, but yeah it should be more readable. I'll change it

File: torch/autograd/functional.py (hunk 9 above)
Comment (zou3519) [reply torch/autograd/functional.py:528]: Oh, I see. Thanks for the clarification, I wasn't thinking being able to vmap through this.

`inputs[el_idx].new_zeros(inputs[el_idx].shape)` and `torch.zeros_like(inputs[el_idx])` are the same with respect to vmap: if `inputs[el_idx]` is a BatchedTensor, then that information gets propagated by `new_zeros` and `zeros_like`. What doesn't propagate information is `torch.zeros` because `torch.zeros` doesn't accept a tensor.

Comment (albanD) [reply torch/autograd/functional.py:528]: Thanks for the clarification!

File: torch/autograd/functional.py (hunk 12 above)
Comment (zou3519) [reply torch/autograd/functional.py:391]: Docstring added to jacobian and hessian docs

**Review by albanD — APPROVED** [review]
Thanks for the update!
